polkadot/runtime/common/src/paras_registrar/mod.rs

```diff
@@ -431,6 +483,46 @@ pub mod pallet {
 			runtime_parachains::set_current_head::<T>(para, new_head);
 			Ok(())
 		}
+
+		/// Changes the stash account of a parachain.
+		///
+		/// The newly set account will be responsible for covering all associated costs related to
+		/// performing parachain validation code upgrades.
+		///
+		/// ## Deposits/Fees
+		/// For this call to be successful, the `new_stash` account must have a sufficient balance
+		/// to cover the current deposit required for this parachain.
+		///
+		/// Can be called by Root, the parachain, or the parachain manager if the parachain is
+		/// unlocked.
+		#[pallet::call_index(9)]
+		#[pallet::weight(<T as Config>::WeightInfo::set_parachain_stash())]
+		pub fn set_parachain_stash(
+			origin: OriginFor<T>,
+			para: ParaId,
+			new_stash: T::AccountId,
+		) -> DispatchResult {
+			Self::ensure_root_para_or_owner(origin, para)?;
```

Comment:
We are going to charge that passed in `new_stash` account. Hence we have to verify that `new_stash` is controlled by the caller. 

We could as a start introduce `set_parachain_billable_stash_to_self` which sets the parachain stash to the origin _iff_ the origin is either the para manager or the parachain itself.

If we wanted more flexibility, we could have instead of `set_parachain_billable_stash_to_self`, two calls:

1. `allow_account_for_parachain_stash(new_stash)` ... where origin needs to be the manager or the parachain itself and merely states, that it would be ok for the passed account `new_stash` to become the stash account.
2. `accept_becoming_billable_stash_for(para_id`). This call will fail if `allow_account_for_parachain_stash` has not been called  for the origin before and the origin is not the manager nor the parachain itself.

`accept_becoming_billable_stash_for` would be identical to `set_parachain_billable_stash_to_self` if the origin is the manager or the parachain itself (no prior `allow_account_for_parachain_stash`) required in that case.

Reply:
On top of that the stash account should be able to revoke itself, in case it no longer wants to be billed for this parachain.

Reply:
Seems like I got really carried away with rewriting the code to the point that I missed such an obvious detail 🙈

I think having `set_parachain_billable_stash_to_self` and `force_set_parachain_billing_account` should be sufficient. 
However, If we want to give more flexibility, we should have a replacing functionality instead of revoking since there must always be an account that has the required reserved.